wallet-new/src/Cardano/Wallet/API/V1/Generic.hs

```diff
@@ -159,4 +157,11 @@ gconsInfos pa = case datatypeInfo pa of
 gconsNames
     :: forall a. (HasDatatypeInfo a, SListI (Code a))
     => Proxy a -> [Text]
-gconsNames = map toText . hcollapse . hliftA gconsName' . gconsInfos
+gconsNames =
+  map toText . hcollapse . hliftA (K . constructorName) . gconsInfos
+
+gconsName
+  :: forall a. (Generic a, HasDatatypeInfo a, SListI (Code a))
+  => a -> Text
+gconsName a =
+  gconsNames (Proxy @a) !! hindex (from a)
```

Comment:
Are you confident partiality (as in using `(!!)`) is not a problem here?

Reply:
I am not yet really familiar with the *generics-sop* module but as far as I understand, this should be fine since we would call it on types represented as a sum (or a singleton) of constructors. As far as I know, we cannot create new types that have no constructor in Haskell? (`newtype` or `data` or extensions that boil down to these two...). Hence, using their corresponding indexes to access constructors' names should work fine; I didn't find any other way to do that with `generics-sop`.

Reply:
With `{-# LANGUAGE EmptyDataDecls  #-}`, you can write `data Void`.

Reply:
I'd like to propose this new theorem:

> Everytime one thinks something isn't possible in Haskell, one discovers a new extension that does it.

Reply:
One either discovers a new extension or patches GHC to add the extension :smile:

Reply:
Though, on the safety benefit, you can't actually *construct* a value of type `Void`, so this function *is* safe, in the sense that `gconsName undefined == undefined` -- the error won't come from `[] !! 0` but from `undefined`!